File: Evaluate Reverse Polish Notation.cpp
class Solution {
public:
    int evalRPN(vector<string> &tokens) {

        int result = 0;
        int i;
        stack<int> opd;         //存储操作数
        int size = tokens.size();
        for(i=0;i<size;i++)
        {
            if(tokens[i]=="*")
            {
                int rOpd = opd.top();   //右操作数
                opd.pop();
                int lOpd = opd.top();  //左操作数
                opd.pop();
                result = lOpd*rOpd;
                opd.push(result);
            }
            else if(tokens[i]=="/")
            {
                int rOpd = opd.top();
                opd.pop();
                int lOpd = opd.top();
                opd.pop();
                result = lOpd/rOpd;
                opd.push(result);
            }
            else if(tokens[i]=="+")
            {
                int rOpd = opd.top();
                opd.pop();
                int lOpd = opd.top();
                opd.pop();
                result = lOpd+rOpd;
                opd.push(result);
            }
            else if(tokens[i]=="-")
            {
                int rOpd = opd.top();
                opd.pop();
                int lOpd = opd.top();
                opd.pop();
                result = lOpd-rOpd;
                opd.push(result);
            }
            else
            {
                opd.push(atoi(tokens[i].c_str()));
            }
        }
        return opd.top();
    }
};

File: Evaluate Reverse Polish Notation.py
import math

class Solution:
    # @param tokens, a list of string
    # @return an integer
    def evalRPN(self, tokens):
        stack = []
        for i in tokens:
            if i == '+':
                x = stack.pop()
                y = stack.pop()
                stack.append(y + x)
                # print("%d %s %d = %d" % (y,i,x,stack[-1]))
            elif i == '-':
                x = stack.pop()
                y = stack.pop()
                stack.append(y - x)
                # print("%d %s %d = %d" % (y,i,x,stack[-1]))
            elif i == '*':
                x = stack.pop()
                y = stack.pop()
                stack.append(y * x)
                # print("%d %s %d = %d" % (y,i,x,stack[-1]))
            elif i == '/':
                x = stack.pop()
                y = stack.pop()
                r = y / x
                if r >= 0:
                    stack.append(int(math.floor(r)))
                else:
                    stack.append(int(math.ceil(r)))
                    # print("%d %s %d = %d" % (y,i,x,stack[-1]))
            else:
                stack.append(int(i))
        return stack.pop()

lst = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]

print((Solution().evalRPN(lst)))
